cmake/CompilerWarnings.cmake

```diff
@@ -0,0 +1,89 @@
+# from here:
+#
+# https://github.com/lefticus/cppbestpractices/blob/master/02-Use_the_Tools_Available.md
+
+# Helper function to enable compiler warnings for a specific set of files
+function(set_file_warnings)
```

Comment:
is there a reason we set the options on a set of files rather than on the targets?

Reply:
We can select precisely for which files we want to enable our strict warnings (which are treated as errors and cause building to fail). There might be cases where we choose to compile some third party source as part of a target and don't want to enable our warnings for upstream code.